handlers.go

```diff
@@ -0,0 +1,389 @@
+	f, err := os.Open(filename)
+	if err != nil {
+		log.WithFields(log.Fields{"errno": FileNotFound}).Warnf(DescribeErrno(FileNotFound), "version.json", err)
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	stat, err := f.Stat()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	http.ServeContent(w, req, "__version__", stat.ModTime(), f)
+}
+
+// Returns a list of known violations for debugging
+func ListViolationsHandler(w http.ResponseWriter, req *http.Request) {
+	val := req.Context().Value(ctxPenaltiesKey)
+	if val == nil {
+		log.WithFields(log.Fields{"errno": RequestContextMissingViolations}).Warnf(DescribeErrno(RequestContextMissingViolations))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	violationPenalties := val.(map[string]uint)
+
+	json, err := json.Marshal(violationPenalties)
+	if err != nil {
+		log.WithFields(log.Fields{"errno": JSONMarshalError}).Warnf(DescribeErrno(JSONMarshalError), "violations", err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(http.StatusOK)
+	w.Write(json)
+}
+
+// UpsertReputationByViolation takes a JSON body from the http request
+// and upserts the reputation entry on the database to the reputation
+// given in reputation violation.  The HTTP requests path has to
+// contain the IP to be updated, in CIDR notation. For example:
+// {"Violation": "password-reset-rate-limit-exceeded"}
+func UpsertReputationByViolationHandler(w http.ResponseWriter, r *http.Request) {
+	splitPath := strings.Split(r.URL.Path, "/")
+	if len(splitPath) != 3 {
+		log.WithFields(log.Fields{"errno": InvalidIPError}).Infof(DescribeErrno(InvalidIPError), r.URL.Path)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	ip, err := IPAddressFromHTTPPath("/" + splitPath[2])
+
+	if err != nil {
+		log.WithFields(log.Fields{"errno": MissingIPError}).Infof(DescribeErrno(MissingIPError), r.URL.Path, err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if !IsValidReputationCIDROrIP(ip) {
+		log.WithFields(log.Fields{"errno": InvalidIPError}).Infof(DescribeErrno(InvalidIPError), splitPath[2])
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.WithFields(log.Fields{"errno": BodyReadError}).Warnf(DescribeErrno(BodyReadError))
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	type ViolationBody struct {
+		Violation string
+	}
+	var entry ViolationBody
+	err = json.Unmarshal(body, &entry)
+	if err != nil {
+		log.WithFields(log.Fields{"errno": JSONUnmarshalError}).Warnf(DescribeErrno(JSONUnmarshalError), err)
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if !IsValidViolationName(entry.Violation) {
+		log.WithFields(log.Fields{"errno": InvalidViolationTypeError}).Infof("Received invalid reputation type: %s", entry.Violation)
```

Comment:
s/reputation/violation

(I already pointed this out somewhere else, but another commit moved it)